### PR TITLE
WiimoteEmu: Implement 3rd-Party extension encryption and uDraw GameTablet.

### PR DESCRIPTION
**Pull request by jordan-woyak**
The tablet uses an alternative "3rd-party" extension encryption which I've implemented.

The Drums, Guitar, and Turntable all seem to use the "1st-party" encryption for some reason.

Now that the 3rd-party encryption is in place adding an emulated TaTaCon will be easy.

### CODE REVIEW

File: Source/Core/Core/HW/WiimoteEmu/Encryption.cpp [hunk 1]
```diff
@@ -6,281 +6,503 @@
...
+        0x4c, 0x93, 0x15, 0xbd, 0xe6, 0x78, 0x03, 0x05, 0x32, 0xf7, 0x34, 0xe4, 0xad, 0x30, 0x98,
+        0xa1, 0x27, 0x9e, 0xf6, 0x43, 0x64, 0x3c, 0x49, 0xbc, 0xf0, 0x99, 0xdd, 0x12, 0xd1, 0xc9,
+        0x5d, 0x7c, 0x22, 0x75, 0x1f, 0x28, 0x2f, 0x09, 0xc2, 0x84, 0x6d, 0xfa, 0x23, 0x3f, 0x7d,
+        0x91, 0x89, 0x82, 0x86, 0x58, 0xdf, 0x76, 0xe9, 0xaa, 0xdb, 0xac, 0x2b, 0xf5, 0x90, 0x42,
+        0x20, 0xe1, 0xa0, 0xed, 0xd4, 0x52, 0x10, 0xef, 0x3e, 0xe3, 0x1c, 0xda, 0x96, 0xb0, 0x72,
+        0x38,
     }};
 
-static inline u8 ROR8(const u8 a, const u8 b)
+constexpr u8 ror8(u8 a, u8 b)
```
Comment (lioncash) [inline Source/Core/Core/HW/WiimoteEmu/Encryption.cpp]: why the lowercasing?

File: Source/Core/Core/HW/WiimoteEmu/Encryption.cpp [hunk 2]
```diff
@@ -6,281 +6,503 @@
...
 {
-// Generate key from the 0x40-0x4c data in g_RegExt
-void EncryptionKey::Generate(const u8* const keydata)
-{
-  u8 rand[10];
-  u8 skey[6];
-  u8 testkey[6];
-  int idx;
+  std::array<u8, 10> rand;
+  for (int i = 0; i != 10; ++i)
```
Comment (lioncash) [inline Source/Core/Core/HW/WiimoteEmu/Encryption.cpp]:
```suggestion
  for (size_t i = 0; i < rand.size(); ++i)
```

File: Source/Core/Core/HW/WiimoteEmu/Encryption.cpp [hunk 3]
```diff
@@ -6,281 +6,503 @@
...
+    rand[9 - i] = key_data[i];
 
-  for (int i = 0; i < 10; ++i)
-    rand[9 - i] = keydata[i];
-  for (int i = 0; i < 6; ++i)
-    skey[5 - i] = keydata[i + 10];
+  std::array<u8, 6> skey;
+  for (int i = 0; i != 6; ++i)
```
Comment (lioncash) [inline Source/Core/Core/HW/WiimoteEmu/Encryption.cpp]:
```suggestion
  for (size_t i = 0; i < skey.size(); ++i)
```

File: Source/Core/Core/HW/WiimoteEmu/Encryption.cpp [hunk 4]
```diff
@@ -6,281 +6,503 @@
...
+  constexpr u8 ans_tbl[7][6] = {
+      {0xa8, 0x77, 0xa6, 0xe0, 0xf7, 0x43}, {0x5a, 0x35, 0x85, 0xe2, 0x72, 0x97},
+      {0x8f, 0xb7, 0x1a, 0x62, 0x87, 0x38}, {0x0d, 0x67, 0xc7, 0xbe, 0x4f, 0x3e},
+      {0x20, 0x76, 0x37, 0x8f, 0x68, 0xb7}, {0xa9, 0x26, 0x3f, 0x2b, 0x10, 0xe3},
+      {0x30, 0x7e, 0x90, 0x0e, 0x85, 0x0a},
+  };
+
+  auto& ans = ans_tbl[idx];
+
+  u8 t0[10];
```
Comment (lioncash) [inline Source/Core/Core/HW/WiimoteEmu/Encryption.cpp]:
```suggestion
  std::array<u8, 10> t0;
```
which allows using `.size()` below to only code the size in one place

File: Source/Core/Core/HW/WiimoteEmu/Encryption.cpp [hunk 5]
```diff
@@ -6,281 +6,503 @@
...
+      u8((ror8((ans[4] ^ t0[1]), t0[6]) - t0[3]) ^ t0[4]),
+      u8((ror8((ans[5] ^ t0[7]), t0[8]) - t0[5]) ^ t0[9]),
+  };
+}
+
+std::array<u8, 6> KeyGen3rdParty::GenerateKey(const std::array<u8, 10>& rand, const u8 idx) const
+{
+  auto& sbox = sboxes[9];
+
+  u8 t0[10];
```
Comment (lioncash) [inline Source/Core/Core/HW/WiimoteEmu/Encryption.cpp]: Ditto, wrt std::array

File: Source/Core/Core/HW/WiimoteEmu/Extension/UDrawTablet.cpp [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  return buttons != 0;
+}
+
+ControllerEmu::ControlGroup* UDrawTablet::GetGroup(UDrawTabletGroup group)
+{
+  switch (group)
+  {
+  case UDrawTabletGroup::Buttons:
+    return m_buttons;
+    break;
```
Comment (lioncash) [inline Source/Core/Core/HW/WiimoteEmu/Extension/UDrawTablet.cpp]: We generally omit break's with returns. These can cause dead code warnings (as well as with various static analysis tooling).

File: Source/Core/Core/HW/WiimoteEmu/Extension/Drums.h [hunk 7]
```diff
@@ -22,8 +22,8 @@ enum class DrumsGroup
   Stick
 };
 
-// TODO: Do the drums ever use encryption?
-class Drums : public EncryptedExtension
+// The Guitar uses the "1st-party" extension encryption scheme.
```
Comment (Warepire) [inline Source/Core/Core/HW/WiimoteEmu/Extension/Drums.h]:
```suggestion
// The Drums uses the "1st-party" extension encryption scheme.
```

File: Source/Core/Core/HW/WiimoteEmu/Encryption.h [hunk 8]
```diff
@@ -6,21 +6,52 @@
...
+{
+public:
+  virtual ~KeyGen() = default;
 
-  void Encrypt(u8* data, int addr, u8 len) const;
-  void Decrypt(u8* data, int addr, u8 len) const;
+  EncryptionKey GenerateFromExtensionKeyData(const std::array<u8, 0x10>& key_data) const;
 
+protected:
+  virtual std::array<u8, 6> GenerateKey(const std::array<u8, 10>& rand, u32 idx) const = 0;
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/WiimoteEmu/Encryption.h]: Thoughts on using a type alias for `std::array<u8, 10>` and `std::array<u8, 6>` for readability?

File: Source/Core/Core/HW/WiimoteEmu/Encryption.h [hunk 9]
```diff
@@ -6,21 +6,52 @@
...
+};
+
+class KeyGen
+{
+public:
+  virtual ~KeyGen() = default;
 
-  void Encrypt(u8* data, int addr, u8 len) const;
-  void Decrypt(u8* data, int addr, u8 len) const;
+  EncryptionKey GenerateFromExtensionKeyData(const std::array<u8, 0x10>& key_data) const;
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/WiimoteEmu/Encryption.h]: It looks a bit weird that this one takes hex 10, while all the other functions use dec 10 (altho it is intentional and not a bug, given that `EncryptedExtension::encryption_key_data` is 16 as well).
Thoughts on using `16` instead? Or a type alias, as mentioned below?

File: Source/Core/Core/HW/WiimoteEmu/Encryption.h (hunk 8 above)
Comment (jordan-woyak) [reply Source/Core/Core/HW/WiimoteEmu/Encryption.h]: :+1:

File: Source/Core/Core/HW/WiimoteEmu/Encryption.h (hunk 9 above)
Comment (jordan-woyak) [reply Source/Core/Core/HW/WiimoteEmu/Encryption.h]: :+1:

**Comment by JMC47** [comment]
image

**Comment by JMC47** [comment]
I can confirm it works

**Pull request merged**